Adds support for asynchronous executions via Promises

acelot opened the pull request:
- New `promiseExecute` method in `Execution.php`
- Helpers for API (`graphqlAsync()`, `processAsync`)

Jalle19 commented:
Would it be possible to use the async version all the time, and just change the `graphql` helper function to wait for the result? That way there would be a lot less code duplication.

acelot commented:
@Jalle19 I think so too, but did so for backward compatibility

Jalle19 commented:
> @Jalle19 I think so too, but did so for backward compatibility

You can think of the `graphql()` helper function and it's related functions as the public API, not the actual public interfaces etc.

acelot commented:
It seems to me that synchronous and asynchronous methods should exist separately, since impossible to turn asynchronous promise into synchronous, because the result of promise comes through a callback.

Jalle19 commented:
Isn't it possible to just add `->then()` and return the result to turn it into synchronous?

acelot commented:
Nope. In any case, will return Promise.

```php
return GraphQL::promiseExecute(...)->then(function (ExecutionResult $result) {
    return $result;
})
```

```php
$result = graphql(...);
var_export($result instanceof Promise); // true
```

Jalle19 commented:
Isn't there some wait method that could be used?

acelot commented:
All promises works in synchronous way without event loop. But thanks to the ReactPHP event loop PHP works in real asynchronous mode. Unfortunately I don't know how ReactPHP resolves promises.

acelot commented:
I have refactored the code. Now `execute` method in `Execution` always returns a Promise. I have moved promise resolving for synchronous calls into API method `graphql`. I also added a check, if a Event Loop is used (React PHP), an error is returned.

acelot commented:
Any chance that this PR will be merged?

Jalle19 commented:
I'll try to find some time to test this out this week, the latest changes look very promising (no pun intended)!

Jalle19 commented:
@acelot I fixed the tests, for some reason Travis CI doesn't run on this pull request. What's the next step in getting deferred resolving to work?

acelot commented:
https://github.com/lordthorzonus/php-dataloader works for me very well. But my project based on ReactPHP. I have not tested data loader in regular PHP environment.

Jalle19 commented:
I still need to test this in a non-asynchronous environment before I dare merge this, I'm not super familiar with part of the code. I have to think if we really want to break the ExecutionInterface signature even though I said it was fine earlier.

crisu83 commented:
Hello @acelot. First of all, I'd like to thank you for all the work you've put into this. I'm going through a divorce so I haven't had any energy to focus on this project for quite some time now, but now I'm back at work after a long vacation so I'm picking it up slowly.

I agree that separate methods is the way to go, and we definitely want to hand over the promise from the public API, instead of resolving it before returning the result. This gives the consumer the freedom to do whatever they need with that promise.

I will now go through the changes and add my comments. After that, if all is good and no more changes are required we can merge this. Thanks once again for your efforts! 😊

crisu83 commented:
We had some problems with deferred resolvers. Any change you've had any success with them @acelot? The problem was that the promises got resolved too early, instead of using a dictionary, delaying the resolution and resolving them all at once.

crisu83 commented on src/api.php:
```diff
@@ -140,6 +142,50 @@ function execute(
+    });
+    if ($data === null) {
+        $data = new ExecutionResult(null, [
+            new GraphQLException('It\'s looks like you are using Event Loop. Please use `executeAsync` method instead.')
```
The message should be: 

```
Looks like you are using Event Loop. Please use `executeAsync` method instead.
```

crisu83 commented on src/api.php:
```diff
@@ -188,7 +234,58 @@ function graphql(
+    });
+    if ($data === null) {
+        $data = new ExecutionResult(null, [
+            new GraphQLException('It\'s looks like you are using Event Loop. Please use `graphqlAsync` method instead.')
```
The message should be: 

```
Looks like you are using Event Loop. Please use `graphqlAsync` method instead.
```

crisu83 commented on src/api.php:
```diff
@@ -188,7 +234,58 @@ function graphql(
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
+    if ($data === null) {
```
I'd flip this condition to `null === $data` for consistency.

crisu83 commented on src/api.php:
```diff
@@ -140,6 +142,50 @@ function execute(
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
+    if ($data === null) {
```
I'd flip this condition to null === $data for consistency.

crisu83 commented on src/api.php:
```diff
@@ -140,6 +142,50 @@ function execute(
+        $errorHandler
+    );
+
+    $data = null;
```
Linebreak after.

crisu83 commented on src/api.php:
```diff
@@ -140,6 +142,50 @@ function execute(
+    $data = null;
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
```
Linebreak after.

crisu83 commented on src/api.php:
```diff
@@ -188,7 +234,58 @@ function graphql(
+        $errorHandler
+    );
+
+    $data = null;
```
Linebreak after.

crisu83 commented on src/api.php:
```diff
@@ -188,7 +234,58 @@ function graphql(
+    $data = null;
+    $resultPromise->then(function (ExecutionResult $result) use (&$data) {
+        $data = $result;
+    });
```
Linebreak after.
